instrument-starters/opentracing-spring-cloud-reactor-starter/src/main/java/io/opentracing/contrib/spring/cloud/reactor/ReactorTracingAutoConfiguration.java

```diff
@@ -37,27 +45,74 @@
  * @author Csaba Kos
  */
 @Configuration
-@AutoConfigureAfter(TracerAutoConfiguration.class)
 @ConditionalOnBean(Tracer.class)
 @ConditionalOnClass(Hooks.class)
 @ConditionalOnProperty(name = "opentracing.spring.cloud.reactor.enabled", havingValue = "true", matchIfMissing = true)
 public class ReactorTracingAutoConfiguration {
+
+  private static final Logger log = Logger.getLogger(ReactorTracingAutoConfiguration.class.getName());
+
   private static final String EXECUTOR_SERVICE_DECORATOR_KEY = ReactorTracingAutoConfiguration.class.getName();
   private static final String HOOK_KEY = ReactorTracingAutoConfiguration.class.getName();
 
-  @Autowired
-  public ReactorTracingAutoConfiguration(final Tracer tracer) {
-    Hooks.onEachOperator(HOOK_KEY, TracedSubscriber.asOperator(tracer));
-    Schedulers.setExecutorServiceDecorator(
-        EXECUTOR_SERVICE_DECORATOR_KEY,
-        (scheduler, scheduledExecutorService) ->
-            new TracedScheduledExecutorService(scheduledExecutorService, tracer)
-    );
+  @Bean
+  static HookRegisteringPostProcessor hookRegisteringPostProcessor(ObjectProvider<Tracer> tracerProvider) {
+    return new HookRegisteringPostProcessor(tracerProvider);
   }
 
-  @PreDestroy
-  public void cleanupHooks() {
-    Hooks.resetOnEachOperator(HOOK_KEY);
-    Schedulers.removeExecutorServiceDecorator(EXECUTOR_SERVICE_DECORATOR_KEY);
+  // Use a BeanDefinitionRegistryPostProcessor here to decorate Reactor as soon as possible before potential early initialization of Reactor schedulers
+  // For more detailed description, please refer to https://github.com/opentracing-contrib/java-spring-cloud/issues/214
+  private static class HookRegisteringPostProcessor implements BeanDefinitionRegistryPostProcessor {
+
+    private final ObjectProvider<Tracer> tracerProvider;
+
+    private volatile Function<? super Publisher<Object>, ? extends Publisher<Object>> hookFunction;
+
+    private HookRegisteringPostProcessor(ObjectProvider<Tracer> tracerProvider) {
+      this.tracerProvider = tracerProvider;
+    }
+
+    @Override
+    public void postProcessBeanDefinitionRegistry(BeanDefinitionRegistry registry) throws BeansException {
+    }
+
+    @Override
+    public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
+      Hooks.onEachOperator(HOOK_KEY, publisher -> {
+        try {
+          return getHookFunction().apply(publisher);
+        } catch (Exception e) {
+          log.log(SEVERE, "Encountered error while retrieving Tracer instance! Fallback to original publisher without hook function..", e);
```

Comment:
This is a good but maybe a more helpful error message for a user of the library would be something like:
`Encountered error while retrieving Tracer instance! This reactive operation chain will not be instrumented.`

(And similarly for the executor service below.)

Reply:
I agree!